src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+
+import { LegacyServiceSetup } from '../legacy';
+import { Env } from '../config';
+import { InternalHttpServiceSetup, KibanaRequest, KibanaResponseFactory } from '../http';
+import { PluginsServiceSetup } from '../plugins';
+import { createCspRuleString } from '../csp';
+import { mergeCapabilities } from './capabilities';
+import { mergeVariables } from './merge_variables';
+import { IApplicationContextProvider } from './types';
+
+interface Params {
+  request: KibanaRequest;
+  response: KibanaResponseFactory;
+  env: Env;
+  legacy: LegacyServiceSetup;
+  http: InternalHttpServiceSetup;
+  uiSettings: Record<string, any>;
+  plugins: PluginsServiceSetup;
+}
+
+interface PluginSpec {
+  readConfigValue: (config: any, key: any) => any;
+}
+
+interface Provider<T = object> {
+  fn: (
+    server: InternalHttpServiceSetup['server'],
+    config: ReturnType<PluginSpec['readConfigValue']>
+  ) => T;
+  pluginSpec: PluginSpec;
+}
+
+type App = ReturnType<InternalHttpServiceSetup['server']['getUiAppById']>;
+type Replacer<T = object> = (
+  variables: T,
+  request: KibanaRequest,
+  server: InternalHttpServiceSetup['server']
+) => Promise<T> | T;
+
+export class ApplicationContextProvider implements IApplicationContextProvider {
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly legacy!: LegacyServiceSetup;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: Record<string, any>;
+  private readonly plugins!: PluginsServiceSetup;
+  private readonly template = compileFile(join(__dirname, 'views/ui_app.pug'));
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
+
+  private getCapabilities() {
+    const modifiers = this.http.server.getCapabilitiesModifiers();
+    const defaultCapabilties = this.http.server.getDefaultCapabilities();
+    const capabilities = mergeCapabilities(defaultCapabilties, {
+      // Get legacy nav links
+      navLinks: Object.assign(
+        {},
+        ...this.http.server.getUiNavLinks().map(({ _id }) => ({ [_id]: true }))
+      ),
+    });
+
+    return modifiers.reduce(
+      async (resolvedCapabilties, modifier) => modifier(this.request, await resolvedCapabilties),
+      Promise.resolve(capabilities)
+    );
+  }
+
+  private async getVariables(app: App, basePath: string, injectedOverrides: Record<string, any>) {
+    const providers: Provider[] = (this.legacy.uiExports as any).defaultInjectedVarProviders || [];
+    const replacers: Replacer[] = (this.legacy.uiExports as any).injectedVarsReplacers || [];
+    const defaultInjectedVars = providers.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(
+          defaults,
+          fn(this.http.server, pluginSpec.readConfigValue(this.http.server.config, []))
+        ),
+      {}
+    );
+    const appInjectedVars = await this.http.server.getInjectedUiAppVars(app.getId());
```

Comment:
Instead of duplicating all that logic in NP, I'd suggest that LP exposed a method to get all required data. Or even keep the rendering in the legacy platform for a while. In this case, we don't need to introduce this unnecessary code, test it and track when it can be removed.